openmeter/productcatalog/plan/adapter/plan.go

```diff
@@ -0,0 +1,537 @@
+	plandb "github.com/openmeterio/openmeter/openmeter/ent/db/plan"
+	phasedb "github.com/openmeterio/openmeter/openmeter/ent/db/planphase"
+	"github.com/openmeterio/openmeter/openmeter/ent/db/predicate"
+	"github.com/openmeterio/openmeter/openmeter/productcatalog/plan"
+	"github.com/openmeterio/openmeter/pkg/datex"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+	"github.com/openmeterio/openmeter/pkg/models"
+	"github.com/openmeterio/openmeter/pkg/pagination"
+	"github.com/openmeterio/openmeter/pkg/sortx"
+)
+
+func (a *adapter) ListPlans(ctx context.Context, params plan.ListPlansInput) (pagination.PagedResponse[plan.Plan], error) {
+	if err := params.Validate(); err != nil {
+		return pagination.PagedResponse[plan.Plan]{}, fmt.Errorf("invalid list Plans parameters: %w", err)
+	}
+
+	query := a.db.Plan.Query()
+
+	if len(params.Namespaces) > 0 {
+		query = query.Where(plandb.NamespaceIn(params.Namespaces...))
+	}
+
+	var orFilters []predicate.Plan
+	if len(params.IDs) > 0 {
+		orFilters = append(orFilters, plandb.IDIn(params.IDs...))
+	}
+
+	if len(params.Keys) > 0 {
+		orFilters = append(orFilters, plandb.KeyIn(params.Keys...))
+	}
+
+	if len(params.KeyVersions) > 0 {
+		for key, version := range params.KeyVersions {
+			orFilters = append(orFilters, plandb.And(plandb.Key(key), plandb.VersionIn(version...)))
+		}
+	}
+
+	query = query.Where(plandb.Or(orFilters...))
+
+	if !params.IncludeDeleted {
+		query = query.Where(plandb.DeletedAtIsNil())
+	}
+
+	// Eager load phases with
+	// * ordering by StartAfter
+	// * with eager load RateCards
+	query = query.WithPhases(
+		planPhaseEagerLoadRateCardsFn,
+	)
+
+	order := entutils.GetOrdering(sortx.OrderDefault)
+	if !params.Order.IsDefaultValue() {
+		order = entutils.GetOrdering(params.Order)
+	}
+
+	switch params.OrderBy {
+	case plan.OrderByCreatedAt:
+		query = query.Order(plandb.ByCreatedAt(order...))
+	case plan.OrderByUpdatedAt:
+		query = query.Order(plandb.ByUpdatedAt(order...))
+	case plan.OrderByVersion:
+		query = query.Order(plandb.ByVersion(order...))
+	case plan.OrderByKey:
+		query = query.Order(plandb.ByVersion(order...))
+	case plan.OrderByID:
+		fallthrough
+	default:
+		query = query.Order(plandb.ByID(order...))
+	}
+
+	response := pagination.PagedResponse[plan.Plan]{
+		Page: params.Page,
+	}
+
+	paged, err := query.Paginate(ctx, params.Page)
+	if err != nil {
+		return response, fmt.Errorf("failed to list Plans: %w", err)
+	}
+
+	result := make([]plan.Plan, 0, len(paged.Items))
+	for _, item := range paged.Items {
+		if item == nil {
+			a.logger.Warn("invalid query result: nil Plan received")
+			continue
+		}
+
+		p, err := fromPlanRow(*item)
+		if err != nil {
+			return response, fmt.Errorf("failed to cast Plan: %w", err)
+		}
+
+		result = append(result, *p)
+	}
+
+	response.TotalCount = paged.TotalCount
+	response.Items = result
+
+	return response, nil
+}
+
+func (a *adapter) CreatePlan(ctx context.Context, params plan.CreatePlanInput) (*plan.Plan, error) {
```

Comment:
I suppose it's already on the way but lets use transaction utils for all repo methods